# Exercise 2

**Author: Andreas Nugroho R**

## Import and load data

Data is imported and loaded. And then from the data, random sampling is applied to get 80% of the data.

In [4]:
import numpy as np
import pandas as pd

In [3]:
df = pd.read_excel('data loan.xlsx')

In [5]:
df.head()

Kontrak Tipe_Kendaraan   Pekerjaan  Area  Tenor Bucket_Collectability  \
0        1         OTHERS    Karyawan  Jawa     12        >90 - 120 hari   
1        2         OTHERS  Wiraswasta  Jawa     12           0 - 30 hari   
2        3         OTHERS  Wiraswasta  Jawa     12           0 - 30 hari   
3        4         OTHERS    Karyawan  Jawa     12           0 - 30 hari   
4        5         OTHERS    Karyawan  Jawa     12           0 - 30 hari   

     NPF  Total_AR     LTV   OTR  ...   DP    AF   Tujuan  Pendidikan  \
0     Ya  2.763126  100.00  2.76  ...  0.0  2.76  Private     Sarjana   
1  Tidak  2.805000  103.70  2.71  ...  0.0  2.81  Private      Others   
2  Tidak  3.104000  103.33  3.00  ...  0.0  3.10  Private     Sarjana   
3  Tidak  4.088100  102.51  3.98  ...  0.0  4.08  Private     Sarjana   
4  Tidak  4.607100  102.22  4.50  ...  0.0  4.60  Private     Sarjana   

   Tanggungan   Status     Gender Paket Inflasi              Segmen  
0   Tidak ada   OTHERS  Laki-laki    Ya    3.58  Jasa & Perdagangan  
1   Tidak ada  MENIKAH  Laki-laki    Ya    3.58              Others  
2   Tidak ada   OTHERS  Perempuan    Ya    4.17              Others  
3         Ada   OTHERS  Laki-laki    Ya    3.61  Jasa & Perdagangan  
4   Tidak ada   OTHERS  Perempuan    Ya    3.30  Jasa & Perdagangan  

[5 rows x 23 columns]

In [6]:
df.shape

(28293, 23)

In [38]:
df_sample = pd.DataFrame.sample(df, frac=0.8)
df_sample.head()

Kontrak   Tipe_Kendaraan   Pekerjaan      Area  Tenor  \
21486    21487  TRUCK & PICK UP  Wiraswasta      Jawa     48   
11269    11270          MINIBUS     Lainnya      Jawa     48   
1193      1194             LCGC  Wiraswasta  Sumatera     36   
20341    20342          MINIBUS  Wiraswasta    Others     60   
10341    10342          MINIBUS     Lainnya  Sumatera     36   

      Bucket_Collectability    NPF    Total_AR    LTV     OTR  ...     DP  \
21486           0 - 30 hari  Tidak  251.952000  85.62  222.47  ...  47.47   
11269           0 - 30 hari  Tidak  156.429993  73.03  173.30  ...  64.54   
1193        >120 - 150 hari     Ya  148.595000  84.66  128.35  ...  32.09   
20341         >30 - 90 hari  Tidak  234.179999  89.67  186.20  ...  46.55   
10341           0 - 30 hari  Tidak  147.235945  68.24  184.50  ...  74.11   

           AF   Tujuan  Pendidikan  Tanggungan   Status     Gender  Paket  \
21486  190.47   Others      Others         Ada  MENIKAH  Laki-laki  Tidak   
11269  126.56  Private      Others   Tidak ada  MENIKAH  Perempuan  Tidak   
1193   108.66  Private     Sarjana   Tidak ada   OTHERS  Perempuan     Ya   
20341  166.96   Others      Others         Ada   OTHERS  Laki-laki  Tidak   
10341  125.90   Others     Sarjana   Tidak ada  MENIKAH  Laki-laki  Tidak   

      Inflasi              Segmen  
21486    3.61  Jasa & Perdagangan  
11269    8.79              Others  
1193     6.29              Others  
20341    8.22  Jasa & Perdagangan  
10341    4.83  Jasa & Perdagangan  

[5 rows x 23 columns]

In [46]:
df_sample.columns

Index(['Kontrak', 'Tipe_Kendaraan', 'Pekerjaan', 'Area', 'Tenor',
       'Bucket_Collectability', 'NPF', 'Total_AR', 'LTV', 'OTR', 'Age',
       'Interest_Rate', 'Income', 'DP', 'AF', 'Tujuan', 'Pendidikan',
       'Tanggungan', 'Status', 'Gender', 'Paket', 'Inflasi', 'Segmen'],
      dtype='object')

## Analysis

### Data Exploration

Some brief data explorations to better understand the data.

In [48]:
df_sample.describe()

Kontrak         Tenor      Total_AR           LTV           OTR  \
count  22634.000000  22634.000000  22634.000000  22634.000000  22634.000000   
mean   14116.886896     45.120085    214.645219     81.118246    204.300174   
std     8168.758574     13.272206     88.348275     14.968910     80.002773   
min        1.000000     12.000000      1.401000      5.110000      1.690000   
25%     7044.750000     36.000000    156.244041     77.830000    157.800000   
50%    14104.500000     48.000000    200.927999     83.820000    187.650000   
75%    21193.750000     60.000000    266.078886     89.390000    230.687500   
max    28291.000000     60.000000   1854.341300    111.830000   1508.400000   

                Age  Interest_Rate        Income            DP            AF  \
count  22634.000000   22634.000000  22634.000000  22634.000000  22634.000000   
mean      40.634046      13.747741     20.280419     59.151714    163.422028   
std        9.558665       3.512262    109.295490     40.818304     64.177110   
min       19.000000       0.000000      1.000000      0.000000      1.390000   
25%       34.000000      11.190000      9.000000     37.740000    122.390000   
50%       40.000000      13.450000     15.000000     47.680000    153.100000   
75%       47.000000      16.210000     20.000000     68.000000    193.050000   
max       82.000000      27.400000  10000.000000    721.350000   1312.810000   

            Inflasi  
count  22634.000000  
mean       5.531018  
std        1.797492  
min        3.210000  
25%        3.830000  
50%        5.310000  
75%        7.250000  
max        8.790000

In [49]:
df_sample['Pekerjaan'].value_counts()

Karyawan      10953
Wiraswasta    10172
Lainnya        1509
Name: Pekerjaan, dtype: int64

In [50]:
df_sample['Bucket_Collectability'].value_counts()

0 - 30 hari        9950
> 150 hari         8476
>90 - 120 hari     1615
>30 - 90 hari      1544
>120 - 150 hari    1049
Name: Bucket_Collectability, dtype: int64

In [51]:
df_sample['Tujuan'].value_counts()

Private    19892
Others      2742
Name: Tujuan, dtype: int64

In [52]:
df_sample['Pendidikan'].value_counts()

Others     12852
Sarjana     9782
Name: Pendidikan, dtype: int64

In [53]:
df_sample['Segmen'].value_counts()

Jasa & Perdagangan    12798
Others                 9836
Name: Segmen, dtype: int64

### Dummy Variables

#### Pekerjaan

In [63]:
d_pekerjaan = pd.get_dummies(df_sample['Pekerjaan'], prefix = 'Kerja')

In [69]:
d_pekerjaan.head(3)

Kerja_Karyawan  Kerja_Lainnya  Kerja_Wiraswasta
21486               0              0                 1
11269               0              1                 0
1193                0              0                 1

In [65]:
d_pekerjaan_others = d_pekerjaan.drop(columns='Kerja_Lainnya')

In [66]:
d_pekerjaan_others.head(3)

Kerja_Karyawan  Kerja_Wiraswasta
21486               0                 1
11269               0                 0
1193                0                 1

In [67]:
df_sample = pd.concat([df_sample, d_pekerjaan_others], axis = 1)

In [68]:
df_sample.head()

Kontrak   Tipe_Kendaraan   Pekerjaan      Area  Tenor  \
21486    21487  TRUCK & PICK UP  Wiraswasta      Jawa     48   
11269    11270          MINIBUS     Lainnya      Jawa     48   
1193      1194             LCGC  Wiraswasta  Sumatera     36   
20341    20342          MINIBUS  Wiraswasta    Others     60   
10341    10342          MINIBUS     Lainnya  Sumatera     36   

      Bucket_Collectability    NPF    Total_AR    LTV     OTR  ...   Tujuan  \
21486           0 - 30 hari  Tidak  251.952000  85.62  222.47  ...   Others   
11269           0 - 30 hari  Tidak  156.429993  73.03  173.30  ...  Private   
1193        >120 - 150 hari     Ya  148.595000  84.66  128.35  ...  Private   
20341         >30 - 90 hari  Tidak  234.179999  89.67  186.20  ...   Others   
10341           0 - 30 hari  Tidak  147.235945  68.24  184.50  ...   Others   

       Pendidikan  Tanggungan   Status     Gender  Paket Inflasi  \
21486      Others         Ada  MENIKAH  Laki-laki  Tidak    3.61   
11269      Others   Tidak ada  MENIKAH  Perempuan  Tidak    8.79   
1193      Sarjana   Tidak ada   OTHERS  Perempuan     Ya    6.29   
20341      Others         Ada   OTHERS  Laki-laki  Tidak    8.22   
10341     Sarjana   Tidak ada  MENIKAH  Laki-laki  Tidak    4.83   

                   Segmen Kerja_Karyawan Kerja_Wiraswasta  
21486  Jasa & Perdagangan              0                1  
11269              Others              0                0  
1193               Others              0                1  
20341  Jasa & Perdagangan              0                1  
10341  Jasa & Perdagangan              0                0  

[5 rows x 25 columns]

#### Pendidikan

In [70]:
d_pendidikan = pd.get_dummies(df_sample['Pendidikan'], prefix = 'Pendidikan')

In [71]:
d_pendidikan.head()

Pendidikan_Others  Pendidikan_Sarjana
21486                  1                   0
11269                  1                   0
1193                   0                   1
20341                  1                   0
10341                  0                   1

In [72]:
d_pendidikan_others = d_pendidikan.drop(columns='Pendidikan_Others')

In [73]:
d_pendidikan_others.head(3)

Pendidikan_Sarjana
21486                   0
11269                   0
1193                    1

In [74]:
df_sample = pd.concat([df_sample, d_pendidikan_others], axis = 1)

In [75]:
df_sample.head(3)

Kontrak   Tipe_Kendaraan   Pekerjaan      Area  Tenor  \
21486    21487  TRUCK & PICK UP  Wiraswasta      Jawa     48   
11269    11270          MINIBUS     Lainnya      Jawa     48   
1193      1194             LCGC  Wiraswasta  Sumatera     36   

      Bucket_Collectability    NPF    Total_AR    LTV     OTR  ...  \
21486           0 - 30 hari  Tidak  251.952000  85.62  222.47  ...   
11269           0 - 30 hari  Tidak  156.429993  73.03  173.30  ...   
1193        >120 - 150 hari     Ya  148.595000  84.66  128.35  ...   

       Pendidikan  Tanggungan   Status     Gender  Paket Inflasi  \
21486      Others         Ada  MENIKAH  Laki-laki  Tidak    3.61   
11269      Others   Tidak ada  MENIKAH  Perempuan  Tidak    8.79   
1193      Sarjana   Tidak ada   OTHERS  Perempuan     Ya    6.29   

                   Segmen Kerja_Karyawan Kerja_Wiraswasta Pendidikan_Sarjana  
21486  Jasa & Perdagangan              0                1                  0  
11269              Others              0                0                  0  
1193               Others              0                1                  1  

[3 rows x 26 columns]

#### NPF

In [107]:
d_npf = pd.get_dummies(df_sample['NPF'], prefix = 'NPF')

In [108]:
d_npf.head()

NPF_Tidak  NPF_Ya
21486          1       0
11269          1       0
1193           0       1
20341          1       0
10341          1       0

In [109]:
d_npf_tidak = d_npf.drop(columns='NPF_Tidak')

In [110]:
df_sample = pd.concat([df_sample, d_npf_tidak], axis = 1)

In [111]:
df_sample.head(3)

Kontrak   Tipe_Kendaraan   Pekerjaan      Area  Tenor  \
21486    21487  TRUCK & PICK UP  Wiraswasta      Jawa     48   
11269    11270          MINIBUS     Lainnya      Jawa     48   
1193      1194             LCGC  Wiraswasta  Sumatera     36   

      Bucket_Collectability    NPF    Total_AR    LTV     OTR  ...  \
21486           0 - 30 hari  Tidak  251.952000  85.62  222.47  ...   
11269           0 - 30 hari  Tidak  156.429993  73.03  173.30  ...   
1193        >120 - 150 hari     Ya  148.595000  84.66  128.35  ...   

       Tanggungan   Status     Gender  Paket  Inflasi              Segmen  \
21486         Ada  MENIKAH  Laki-laki  Tidak     3.61  Jasa & Perdagangan   
11269   Tidak ada  MENIKAH  Perempuan  Tidak     8.79              Others   
1193    Tidak ada   OTHERS  Perempuan     Ya     6.29              Others   

      Kerja_Karyawan Kerja_Wiraswasta Pendidikan_Sarjana NPF_Ya  
21486              0                1                  0      0  
11269              0                0                  0      0  
1193               0                1                  1      1  

[3 rows x 27 columns]

### Hypothesis Testing

Apakah rata-rata pendidikan sarjana pada kelompok perempuan berbeda dari pendidikan sarjana pada kelompok laki-laki?

In [76]:
from scipy.stats import ttest_ind

In [98]:
x1 = df_sample['Pendidikan_Sarjana'][df_sample['Gender'] == 'Perempuan']
x2 = df_sample['Pendidikan_Sarjana'][df_sample['Gender'] != 'Perempuan']

In [99]:
ttest_ind(x1, x2, equal_var = True) # assuming equal variances

Ttest_indResult(statistic=10.937755959519487, pvalue=8.927148714123158e-28)

In [100]:
print('standard deviations of x1 = %.3f and x2 = %.3f' %(x1.std(), x2.std()))

standard deviations of x1 = 0.500 and x2 = 0.492


**Interpretasi** <br>
Karena p-value > 0.05 dan deviasi standar antara x1 dan x2 yang tidak terlalu bebeda, maka H0 gagal ditolak. Sehingga dapat disimpulkan bahwa rata-rata pendidikan sarjana pada kelompok perempuan dan laki-laki tidak berbeda secara signifikan.

Apakah rata-rata pekerjaan wiraswasta pada kelompok NPF berbeda dari pekerjaan wiraswasta pada kelompok non-NPF?

In [103]:
a1 = df_sample['Kerja_Wiraswasta'][df_sample['NPF'] == 'Ya']
a2 = df_sample['Kerja_Wiraswasta'][df_sample['NPF'] != 'Ya']

In [104]:
ttest_ind(a1, a2, equal_var = True) # assuming equal variances

Ttest_indResult(statistic=16.03408884368722, pvalue=1.5330532485909583e-57)

In [106]:
print('standard deviations of a1 = %.3f and a2 = %.3f' %(a1.std(), a2.std()))

standard deviations of a1 = 0.500 and a2 = 0.489


**Interpretasi** <br>
Karena p-value > 0.05 dan deviasi standar antara a1 dan a2 yang tidak terlalu bebeda, maka H0 gagal ditolak. Sehingga dapat disimpulkan bahwa rata-rata pekerjaan wiraswasta pada kelompok NPF dan non-NPF tidak berbeda secara signifikan.

### Linear Regression

In [112]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn import metrics

In [113]:
# input dataset
x = df_sample['NPF_Ya']
y = df_sample['LTV']

In [114]:
# change data structure
x = x.to_numpy().reshape(-1,1)
y = y.to_numpy().reshape(-1,1)

In [115]:
regressor = LinearRegression()
regressor.fit(x,y) # training the algorithm

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [116]:
# regression equation
print('y = %.3f + %.3f X' %(regressor.intercept_, regressor.coef_))

y = 75.553 + 11.307 X


In [117]:
# regression equation
y_pred = regressor.predict(x)
print('R2 = %.3f' %metrics.r2_score(y, y_pred))
rmse = np.sqrt(metrics.mean_squared_error(y, y_pred))
print('RMSE = %.3f' %rmse)

R2 = 0.143
RMSE = 13.860


**Interpretasi**<br>
Model hanya menjelaskan 14.3% variasi data di sekitar rata-ratanya.

### Modelling with statsmodels

**Regresi pengaruh pendapatan, pendidikan sarjana, pekerjaan karyawan, inflasi, dan tenor terhadap kondisi NPF**

In [118]:
import statsmodels.formula.api as smf

In [119]:
ols_df_sample = smf.ols('NPF_Ya ~ Income + Pendidikan_Sarjana + Kerja_Karyawan + Inflasi + Tenor', data=df_sample).fit()
print(ols_df_sample.summary())

                            OLS Regression Results                            
Dep. Variable:                 NPF_Ya   R-squared:                       0.122
Model:                            OLS   Adj. R-squared:                  0.122
Method:                 Least Squares   F-statistic:                     630.3
Date:                Fri, 25 Oct 2019   Prob (F-statistic):               0.00
Time:                        23:02:10   Log-Likelihood:                -14949.
No. Observations:               22634   AIC:                         2.991e+04
Df Residuals:                   22628   BIC:                         2.996e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
Intercept             -0.0386      0